sky/spot/spot_utils.py

```diff
@@ -68,6 +68,41 @@ def get_job_status(backend: 'backends.CloudVmRayBackend',
     return status
 
 
+def update_spot_job_status(job_id: Optional[int] = None):
+    """Update spot job status if the controller failed abnormally.
+
+    Check the status of the controller. If it is not running, it must be
+    exited abnormally, and we should set the job status to FAILED.
+    """
+    if job_id is None:
+        job_ids = spot_state.get_nonterminal_job_ids_by_name(None)
+    else:
+        job_ids = [job_id]
+    for job in job_ids:
+        controller_status = job_lib.get_status(job)
+        if controller_status.is_terminal():
+            logger.error(f'Controller for job {job} have exited abnormally. '
+                         'Setting the job status to FAILED_CONTROLLER.')
+            task_name = spot_state.get_task_name_by_job_id(job)
+
+            # Tear down the abnormal spot cluster to avoid resource leakage.
+            cluster_name = generate_spot_cluster_name(task_name, job)
+            handle = global_user_state.get_handle_from_cluster_name(
+                cluster_name)
+            if handle is not None:
+                backend = backend_utils.get_backend_from_handle(handle)
+                try:
+                    backend.teardown(handle, terminate=True)
+                except RuntimeError:
+                    logger.error('Failed to tear down the spot cluster '
```

Comment:
Should we avoid setting L103 if we hit this termination error? So it's easier to spot potential leakage.

Reply:
My understanding is 
```
  - FAILED_CONTROLLER: The job failed due to an unexpected error in the spot
      controller.
```
suggests the spot job failed, but it could've leaked if termination fails. Should we treat non-terminal statuses = cluster potentially alive, and terminal = cluster definitely down?